Lib/pdb.py

```diff
@@ -2491,6 +2502,465 @@ def set_trace(*, header=None, commands=None):
+            except (KeyboardInterrupt, EOFError):
+                print(flush=True)
+                print("command definition aborted, old commands restored")
+                break
+            finally:
+                self.commands_mode = False
+
+        command = "\n".join(parts)
+        self.sockfile.write((json.dumps({"command": command}) + "\n").encode())
+        self.sockfile.flush()
+
+    def prompt_for_repl_command(self, prompt):
+        while True:
+            try:
+                command = self.read_command(prompt)
+            except EOFError:
+                command = "EOF"
+            except KeyboardInterrupt:
+                print(flush=True)
+                continue
+            except Exception as exc:
+                msg = traceback.format_exception_only(exc)[-1].strip()
+                print("***", msg, flush=True)
+                continue
+
+            if self.interact_mode:
+                payload = {"interact": command}
+            else:
+                payload = {"command": command}
+            self.sockfile.write((json.dumps(payload) + "\n").encode())
+            self.sockfile.flush()
+            return
+
+    def complete(self, text, state):
+        import readline
+
+        if state == 0:
+            origline = readline.get_line_buffer()
+            line = origline.lstrip()
+            stripped = len(origline) - len(line)
+            begidx = readline.get_begidx() - stripped
+            endidx = readline.get_endidx() - stripped
+
+            msg = {
+                "completion": {
+                    "text": text,
+                    "line": line,
+                    "begidx": begidx,
+                    "endidx": endidx,
+                }
+            }
+            self.sockfile.write((json.dumps(msg) + "\n").encode())
+            self.sockfile.flush()
+
+            payload = self.sockfile.readline()
+            payload = json.loads(payload)
+            if "completion" not in payload:
+                raise RuntimeError(
+                    f"Failed to get valid completions. Got: {payload}"
+                )
+
+            self.completion_matches = payload["completion"]
+        try:
+            return self.completion_matches[state]
+        except IndexError:
+            return None
+
+
+def _connect(host, port, frame):
+    with closing(socket.create_connection((host, port))) as conn:
+        sockfile = conn.makefile("rwb")
+
+    remote_pdb = _RemotePdb(sockfile)
+    weakref.finalize(remote_pdb, sockfile.close)
+    remote_pdb.set_trace(frame=frame)
+
+
+def attach(pid):
+    """Attach to a running process with the given PID."""
+    with closing(socket.create_server(("localhost", 0))) as server:
+        port = server.getsockname()[1]
+
+        with tempfile.NamedTemporaryFile("w", delete_on_close=False) as script:
+            script.write(
+                f'import pdb, sys\n'
+                f'pdb._connect("localhost", {port}, sys._getframe().f_back)\n'
```

Comment:
Actually, I prefer to add a one-shot unique string on the protocol(uuid or anything else). The pdb server will deny a request when there is not an auth string in the request or the string is not correct). I think will make the debug process more secure(The malware/virus can not just scan the process and port and collect information from the PDB server).

Reply:
> The malware/virus can not just scan the process and port and collect information from the PDB server

I'm not sure that this makes sense... note that the server here is the thing that's doing the _attaching_, not the process that's being attached _to_. The TCP server is the PDB client, and the PDB server is the TCP client. So the TCP server has nothing secret in it that the attacker could learn by connecting to it: it's literally just the `__main__` of the `pdb` module. All the secrets are held by the TCP client that's running the PDB command loop.

Reply:
> I'm not sure that this makes sense... note that the server here is the thing that's doing the _attaching_, not the process that's being attached _to_. The TCP server is the PDB client, and the PDB server is the TCP client. So the TCP server has nothing secret in it that the attacker could learn by connecting to it: it's literally just the `__main__` of the `pdb` module. All the secrets are held by the TCP client that's running the PDB command loop.

Ops, Thanks for the reply, I misunderstand for the role for `server` and `client`. You are right